libs/core/src/main/java/org/elasticsearch/core/Releasables.java

```diff
@@ -165,4 +151,23 @@ public String toString() {
             return delegate;
         }
     }
+
+    private static class ReleaseOnce extends AtomicReference<Releasable> implements Releasable {
```

Comment:
This seems ok because the caller can't see the `AtomicReference` bit. Maybe do the same thing to `ActionListener#notifyOnce` and `RunOnce` too?

Reply:
Ah for those I can't do without making things visible to the caller I think. At least not unless I move more things around. Maybe do that later if we see it matter anywhere?